### PR TITLE
uart: ns16550: rework device init marcos

### PR DESCRIPTION
**Pull request by dcpleung**
The device init macros relied on too many macro trampolines and it's getting really hard to decode what's going on. So rework the init macros.

### CODE REVIEW

**Comment by najumon1980** [comment]
I also created PR (https://github.com/zephyrproject-rtos/zephyr/pull/62993) for abstract platform complexity from device drivers and updated ns16550 accordingly.

**Review by jhedberg — APPROVED** [review]

File: drivers/serial/uart_ns16550.c [hunk 1]
```diff
@@ -249,7 +264,7 @@ struct uart_ns16550_device_config {
 #if defined(CONFIG_PINCTRL)
 	const struct pinctrl_dev_config *pincfg;
 #endif
-#if defined(CONFIG_UART_NS16550_ACCESS_IOPORT) || defined(CONFIG_UART_NS16550_SIMULT_ACCESS)
+#if UART_NS16550_IOPORT_ENABLED
```
Comment (najumon1980) [inline drivers/serial/uart_ns16550.c:267]: please check some platform still using CONFIG_UART_NS16550_ACCESS_IOPORT ?

Comment (jhedberg) [reply drivers/serial/uart_ns16550.c:267]: > please check some platform still using CONFIG_UART_NS16550_ACCESS_IOPORT ?

I did a quick grep in the tree and couldn't find any remaining users (of course it's possible there are out-of-tree platforms, but we can't do much about those)

**Comment by najumon1980** [comment]
Looks the PR ok for me, will do a quick sanity check on Intel platforms and update you ?

**Comment by najumon1980** [comment]
> Looks the PR ok for me, will do a quick sanity check on Intel platforms and update you ?

Sanity test passed

**Review by najumon1980 — APPROVED** [review]

File: drivers/serial/uart_ns16550.c [hunk 2]
```diff
@@ -1220,10 +1220,10 @@ static const struct uart_driver_api uart_ns16550_driver_api = {
 #endif
 };
 
-#define UART_NS16550_IRQ_FLAGS_SENSE0(n) 0
-#define UART_NS16550_IRQ_FLAGS_SENSE1(n) DT_INST_IRQ(n, sense)
 #define UART_NS16550_IRQ_FLAGS(n) \
-	_CONCAT(UART_NS16550_IRQ_FLAGS_SENSE, DT_INST_IRQ_HAS_CELL(n, sense))(n)
+	COND_CODE_1(DT_INST_IRQ_HAS_CELL(n, sense),                           \
+		    (DT_INST_IRQ(n, sense)),                                  \
+		    (0))
```
Comment (finikorg) [inline drivers/serial/uart_ns16550.c:1242]: like it, maybe create common macro for other drivers to use?

Comment (najumon1980) [reply drivers/serial/uart_ns16550.c:1242]: I already created a PR (https://github.com/zephyrproject-rtos/zephyr/pull/62993) for use common for all drivers

**Comment by finikorg** [comment]
> > Looks the PR ok for me, will do a quick sanity check on Intel platforms and update you ?
> 
> Sanity test passed

Have you tested shell working?

**Pull request merged**